### DISCUSSION TITLE
Budibase DB `oneOf` query issues

### DISCUSSION THREAD

**Discussion by brocococonut** [Budibase/budibase#6419] [Help]
I've been looking for a way to use `OR` queries on Budibase DB tables, but alas - there's no way to. I did stumble across the `oneOf` filter in the API explorer though. Using the API explorer, roughly the following options object gets generated:

```javascript
const options = {
  method: "POST",
  headers: {
    Accept: "application/json",
    "x-budibase-app-id": "....",
    "Content-Type": "application/json",
    "x-budibase-api-key":
      "......",
  },
  body: JSON.stringify({
    query: {
      oneOf: {
        _id: "ro_ta_aed1274ac0024deb9fdd864494bb9266_5103357049514eb2972fbcb4a2d19890",
      },
    },
  }),
};
```

This however returns all rows in the table as if there was no `oneOf` filter applied. Now, the docs for the filter imply that the value for each column key should actually be an `Array`, not a string. So if we alter it to:
```javascript
.......
      oneOf: {
        _id: ["ro_ta_aed1274ac0024deb9fdd864494bb9266_5103357049514eb2972fbcb4a2d19890"],
      },
.......
```
You'd expect that to work... However it's completely ignored like its predecessor. Am I missing anything here? `oneOf` does seem to exist in the code for Budibase, but I can't really tell if it's set up for anything other than `SQL` resources... If it's only for SQL resources, then why would it be in the API explorer for the `/tables/:id/rows/search` route?

You can see some of where it's implemented below:
https://github.com/Budibase/budibase/blob/cb153c7da6ce3f85989f02a2e5817a7fda6f7eed/packages/server/src/constants/index.js#L18-L20
https://github.com/Budibase/budibase/blob/cb153c7da6ce3f85989f02a2e5817a7fda6f7eed/packages/server/src/api/routes/utils/validators.js#L57-L59
https://github.com/Budibase/budibase/blob/cb153c7da6ce3f85989f02a2e5817a7fda6f7eed/packages/server/src/api/controllers/row/ExternalRequest.ts#L127-L135

**Comment by brocococonut** [comment]
Additionally, I've noticed the `range` option for the public API also does nothing...

**Reply by brocococonut** [reply]
Welp, I figured out the `range` filter at least... I had to look at the input to an automation step to see how it's actually meant to be formatted.
In the API explorer, it says the following:
`Searches within a range, the format of this must be columnName -> [low, high].`
When you test it out, it provides the following example:
```javascript
const sdk = require("api")("@budibase/v1.0.10#3iote64ll1c4frc5");

sdk.auth(
  "......."
);
sdk.server(".....");
sdk
  .post("/tables/tableId/rows/search", {
    query: { range: { "Auto ID": "1, 2" } },
  })
  .then((res) => console.log(res))
  .catch((err) => console.error(err));

```
Now, both the generated example and the description of the filter are just flat out wrong. The CORRECT example should be something like:
```javascript
const sdk = require("api")("@budibase/v1.0.10#3iote64ll1c4frc5");

sdk.auth(
  "....."
);
sdk.server("....");
sdk
  .post("/tables/tableId/rows/search", {
    query: {
      range: {
        "Auto ID": {
          low: 1,
          high: 2,
        },
      },
    },
  })
  .then((res) => console.log(res))
  .catch((err) => console.error(err));
```

This should definitely be made more clear - at least somewhere in the docs. I don't think I've missed it at least..

**Reply by brocococonut** [reply]
The above lead me to look into the `oneOf` field another time, and unlike the `range`, it just straight up doesn't exist on the `QueryBuilder` class..
Here's the location of the range code in the `QueryBuilder`, where the second snippet is the only other place I can see the format of `range` described.

https://github.com/Budibase/budibase/blob/9c8e25526e0abd96bc85156f7040e837555e2fdb/packages/server/src/api/controllers/row/internalSearch.js#L14-L16
https://github.com/Budibase/budibase/blob/9c8e25526e0abd96bc85156f7040e837555e2fdb/packages/server/src/api/controllers/row/internalSearch.js#L87-L93

What's missing from the file shown in the above two snippets is the `oneOf` filter

**Comment by mike12345567** [comment]
Hi @brocococonut - thanks for looking into all this, I'll get the `range` documentation updated, its not good when documentation isn't accurate, thanks for raising an issue for it.

As for one of, its sadly a by-product of our single API for all datasources, it is supported for SQL datasources but isn't currently supported by Budibase DB. This should be reflected in the docs as well, as there isn't really an exposed method that would allow us to implement one of right now.

As for using OR operators in the Budibase DB, we know this is a bit of a limitation right now, we do want to expose some sort of method for querying Budibase DB with a selection of boolean operations, but right now that isn't really possible. For SQL databases you can specify an `allOr: true` parameter, which switches everything to using OR, but this currently isn't available for our Lucene index on Budibase DB - this is something we could look into enabling as it might not be too difficult and could be useful for API users.

Sorry that you needed to do such a deep dive to work this out, we should add more documentation around this to the specification, making it clear the difference between the SQL search and the Lucene search for internal tables - for the most part they are quite similar (when using the UI) but under the hood there are quite a few more options for SQL as the search capabilities are much greater.
